core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol4.java

```diff
@@ -175,11 +149,18 @@ public String getDisplayName() {
 
     @Override
     public String getName() {
-        return handler.getName();
+        return "JNLP4-connect"; // matches JnlpProtocol4Handler.getName
     }
 
     @Override
     public void handle(Socket socket) throws IOException, InterruptedException {
+        try {
+            init();
+        } catch (IOException x) {
+            throw x;
+        } catch (Exception x) {
+            throw new IOException(x);
+        }
```

Comment:
```suggestion
        if (handler == null) {
	        try {
	            init();
	        } catch (IOException x) {
	            throw x;
	        } catch (Exception x) {
	            throw new IOException(x);
	        }
        }
```

to avoid taking the object monitor?

Reply:
Would have to think about whether that requires making `handler` be `volatile`. Does not seem necessary since `handle` should be called once per actual agent connection, which is many orders of magnitude more expensive than acquiring an uncontended monitor.

Reply:
FTR [this pattern](https://errorprone.info/bugpattern/DoubleCheckedLocking) would work, but as that page itself mentions, "in modern JVMs with efficient uncontended synchronization the performance difference is often negligible". The performance of the current version is certainly fine as described in the previous comment.